### PR TITLE
Add Unit Tests for User Sharing Component

### PR DESCRIPTION
**Pull request by BimsaraBodaragama**
## Purpose
The User Sharing Component currently lacks unit test coverage, which affects its reliability and maintainability. This PR introduces unit tests to validate key functionalities and ensure expected behavior under different conditions.

## Goals
- Improve test coverage for the User Sharing Component.
- Ensure correctness by validating core functionalities.
- Enhance maintainability and prevent regressions in future updates.

## Approach
- Add unit tests covering various use cases, including validation logic and business rules.
- Ensure test cases handle both expected and edge cases.
- Follow best practices for structuring unit tests to maintain readability and maintainability.

### CODE REVIEW

File: components/org.wso2.carbon.identity.organization.management.organization.user.sharing/src/test/java/org/wso2/carbon/identity/organization/management/organization/user/sharing/UserSharingPolicyHandlerServiceImplTest.java [hunk 1]
```diff
@@ -0,0 +1,479 @@
...
+import static org.wso2.carbon.identity.organization.management.organization.user.sharing.constant.TestUserSharingConstants.VALIDATE_MSG_SHARED_ORG_ID;
+import static org.wso2.carbon.identity.organization.management.organization.user.sharing.constant.TestUserSharingConstants.VALIDATE_MSG_SHARED_ORG_NAME;
+import static org.wso2.carbon.identity.organization.management.organization.user.sharing.constant.TestUserSharingConstants.VALIDATE_MSG_SHARED_ROLE_AUDIENCE_NAME;
+import static org.wso2.carbon.identity.organization.management.organization.user.sharing.constant.TestUserSharingConstants.VALIDATE_MSG_SHARED_ROLE_AUDIENCE_TYPE;
+import static org.wso2.carbon.identity.organization.management.organization.user.sharing.constant.TestUserSharingConstants.VALIDATE_MSG_SHARED_ROLE_NAME;
+import static org.wso2.carbon.identity.organization.management.organization.user.sharing.constant.TestUserSharingConstants.VALIDATE_MSG_SHARED_TYPE;
+import static org.wso2.carbon.identity.organization.management.organization.user.sharing.constant.TestUserSharingConstants.VALIDATE_MSG_SHARED_USER_ID;
```
Comment (pavinduLakshan) [inline components/org.wso2.carbon.identity.organization.management.organization.user.sharing/src/test/java/org/wso2/carbon/identity/organization/management/organization/user/sharing/UserSharingPolicyHandlerServiceImplTest.java:101]: just asking out of curiosity, cannot we import all constants as a whole instead of importing each constant separately?

File: components/org.wso2.carbon.identity.organization.management.organization.user.sharing/src/test/java/org/wso2/carbon/identity/organization/management/organization/user/sharing/UserSharingPolicyHandlerServiceImplTest.java [hunk 2]
```diff
@@ -0,0 +1,479 @@
...
+
+public class UserSharingPolicyHandlerServiceImplTest {
```
Comment (pavinduLakshan) [inline components/org.wso2.carbon.identity.organization.management.organization.user.sharing/src/test/java/org/wso2/carbon/identity/organization/management/organization/user/sharing/UserSharingPolicyHandlerServiceImplTest.java:103]: Missing doc comment for class

File: components/org.wso2.carbon.identity.organization.management.organization.user.sharing/src/test/java/org/wso2/carbon/identity/organization/management/organization/user/sharing/UserSharingPolicyHandlerServiceImplTest.java (hunk 1 above)
Comment (BimsaraBodaragama) [reply components/org.wso2.carbon.identity.organization.management.organization.user.sharing/src/test/java/org/wso2/carbon/identity/organization/management/organization/user/sharing/UserSharingPolicyHandlerServiceImplTest.java:101]: It is not recommended to use wildcard imports `(*)`, as this violates wso2 checkstyle rules and can lead to maintainability issues.

```
[ERROR] /Users/BimsaraB/WSO2-Workspace/Repos/wso2-extensions/identity-organization-management/components/org.wso2.carbon.identity.organization.management.organization.user.sharing/src/test/java/org/wso2/carbon/identity/organization/management/organization/user/sharing/UserSharingPolicyHandlerServiceImplTest.java:61: Using the '.*' form of import should be avoided - org.wso2.carbon.identity.organization.management.organization.user.sharing.constant.TestUserSharingConstants.*. [AvoidStarImport]
```

File: components/org.wso2.carbon.identity.organization.management.organization.user.sharing/src/test/java/org/wso2/carbon/identity/organization/management/organization/user/sharing/UserSharingPolicyHandlerServiceImplTest.java (hunk 2 above)
Comment (BimsaraBodaragama) [reply components/org.wso2.carbon.identity.organization.management.organization.user.sharing/src/test/java/org/wso2/carbon/identity/organization/management/organization/user/sharing/UserSharingPolicyHandlerServiceImplTest.java:106]: Addressed in [c4503af](https://github.com/wso2-extensions/identity-organization-management/pull/486/commits/c4503af5379e9d795c1789c3dfc718dde4ca7205)

File: components/org.wso2.carbon.identity.organization.management.organization.user.sharing/src/test/java/org/wso2/carbon/identity/organization/management/organization/user/sharing/UserSharingPolicyHandlerServiceImplTest.java [hunk 3]
```diff
@@ -0,0 +1,482 @@
...
+                {USER_2_ID, setExpectedResultsForGetSharedOrgsTestCase2()},
+                {USER_3_ID, setExpectedResultsForGetSharedOrgsTestCase3()}
+        };
+    }
+
+    @Test(dataProvider = "getSharedOrgsDataProvider")
+    public void testGetSharedOrganizationsOfUser(String associatedUserId, Map<String, UserAssociation> expectedResults)
+            throws Exception {
+
+        // Mock getOrganizationId() from the Utils class.
```
Comment (SujanSanjula96) [inline components/org.wso2.carbon.identity.organization.management.organization.user.sharing/src/test/java/org/wso2/carbon/identity/organization/management/organization/user/sharing/UserSharingPolicyHandlerServiceImplTest.java:148]: Do we need this comments

Comment (BimsaraBodaragama) [reply components/org.wso2.carbon.identity.organization.management.organization.user.sharing/src/test/java/org/wso2/carbon/identity/organization/management/organization/user/sharing/UserSharingPolicyHandlerServiceImplTest.java]: Addressed in [21ea0a1](https://github.com/wso2-extensions/identity-organization-management/pull/486/commits/21ea0a1f1bf6e77cad927098126e6992783d1210)

File: components/org.wso2.carbon.identity.organization.management.organization.user.sharing/src/test/java/org/wso2/carbon/identity/organization/management/organization/user/sharing/UserSharingPolicyHandlerServiceImplTest.java [hunk 4]
```diff
@@ -0,0 +1,469 @@
...
+        when(dataHolder.getOrganizationUserSharingService()).thenReturn(mockOrgUserSharingService);
+
+        when(mockOrgUserSharingService.getUserAssociationsOfGivenUser(anyString(), anyString())).thenThrow(
+                new OrganizationManagementException(VALIDATE_MSG_EXCEPTION));
+
+        // Invoke the method (expected to throw an exception).
+        userSharingPolicyHandlerService.getSharedOrganizationsOfUser(USER_1_ID, null, null, null, null, false);
+    }
+
+    // Tests for GET Shared Roles.
```
Comment (ShanChathusanda93) [inline components/org.wso2.carbon.identity.organization.management.organization.user.sharing/src/test/java/org/wso2/carbon/identity/organization/management/organization/user/sharing/UserSharingPolicyHandlerServiceImplTest.java:206]: The tests method name should be self explanatory. If you need to add more details you can use the description option for the test method.

File: components/org.wso2.carbon.identity.organization.management.organization.user.sharing/src/test/java/org/wso2/carbon/identity/organization/management/organization/user/sharing/UserSharingPolicyHandlerServiceImplTest.java [hunk 5]
```diff
@@ -0,0 +1,469 @@
...
+
+        Map<String, UserAssociation> expectedResults = new HashMap<>();
+        expectedResults.put(ORG_1_NAME, createUserAssociation(USER_3_ID, ORG_1_ID));
+        return expectedResults;
+    }
+
+    private Map<String, UserAssociation> setExpectedResultsForGetSharedOrgsTestCase3() {
+
+        return new HashMap<>(); // No shared orgs for this test case.
+    }
```
Comment (ShanChathusanda93) [inline components/org.wso2.carbon.identity.organization.management.organization.user.sharing/src/test/java/org/wso2/carbon/identity/organization/management/organization/user/sharing/UserSharingPolicyHandlerServiceImplTest.java:369]: Do we need this private method?

File: components/org.wso2.carbon.identity.organization.management.organization.user.sharing/src/test/java/org/wso2/carbon/identity/organization/management/organization/user/sharing/UserSharingPolicyHandlerServiceImplTest.java [hunk 6]
```diff
@@ -0,0 +1,469 @@
...
+    @DataProvider(name = "roleSharingDataProvider")
+    public Object[][] roleSharingDataProvider() {
+
+        return new Object[][]{
+                {USER_1_ID, ORG_1_ID, setExpectedResultsForGetSharedRolesTestCase1()},
+                {USER_2_ID, ORG_2_ID, setExpectedResultsForGetSharedRolesTestCase2()},
+                {USER_3_ID, ORG_3_ID, setExpectedResultsForGetSharedRolesTestCase3()},
+                {USER_1_ID, ORG_1_ID, setExpectedResultsForGetSharedRolesTestCase4()},
+                {USER_1_ID, ORG_1_ID, setExpectedResultsForGetSharedRolesTestCase5()},
+                {USER_1_ID, ORG_1_ID, setExpectedResultsForGetSharedRolesTestCase6()}
```
Comment (ShanChathusanda93) [inline components/org.wso2.carbon.identity.organization.management.organization.user.sharing/src/test/java/org/wso2/carbon/identity/organization/management/organization/user/sharing/UserSharingPolicyHandlerServiceImplTest.java:217]:
```suggestion
                {USER_1_ID, ORG_1_ID, Collections.emptyList()}
```

File: components/org.wso2.carbon.identity.organization.management.organization.user.sharing/src/test/java/org/wso2/carbon/identity/organization/management/organization/user/sharing/UserSharingPolicyHandlerServiceImplTest.java [hunk 7]
```diff
@@ -0,0 +1,469 @@
...
+
+            UserAssociation userAssociation = createUserAssociation(userId, orgId);
+            when(mockOrgUserSharingService.getUserAssociationOfAssociatedUserByOrgId(userId, orgId)).thenReturn(
+                    userAssociation);
+
+            when(mockOrgManager.resolveTenantDomain(anyString())).thenReturn(TENANT_DOMAIN);
+            identityTenantUtilMockedStatic.when(() -> IdentityTenantUtil.getTenantId(TENANT_DOMAIN))
+                    .thenReturn(TENANT_ID);
+
+            String domainQualifiedUserName = getDomainQualifiedUserName(userId);
```
Comment (ShanChathusanda93) [inline components/org.wso2.carbon.identity.organization.management.organization.user.sharing/src/test/java/org/wso2/carbon/identity/organization/management/organization/user/sharing/UserSharingPolicyHandlerServiceImplTest.java:252]:
```suggestion
            String domainQualifiedUserName = USER_NAME_PREFIX + USER_DOMAIN_PRIMARY + PATH_SEPARATOR + userId;
```

File: components/org.wso2.carbon.identity.organization.management.organization.user.sharing/src/test/java/org/wso2/carbon/identity/organization/management/organization/user/sharing/constant/TestUserSharingConstants.java [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+    public static final String FIELD_USER_ID_RESOLVER = "userIDResolver";
+    public static final String TENANT_DOMAIN = "tenantDomain";
+    public static final int TENANT_ID = -1234;
+    public static final String USER_DOMAIN_PRIMARY = "PRIMARY";
+    public static final String USER_NAME_PREFIX = "username-of-";
+    public static final String PATH_SEPARATOR = "/";
+
+    // Mocks.
+    public static final String MOCKED_DATA_ACCESS_EXCEPTION = "Mocked DataAccessException";
+    public static final String MOCKED_TRANSACTION_EXCEPTION = "Mocked TransactionException";
```
Comment (ShanChathusanda93) [inline components/org.wso2.carbon.identity.organization.management.organization.user.sharing/src/test/java/org/wso2/carbon/identity/organization/management/organization/user/sharing/constant/TestUserSharingConstants.java:36]: Cannot see any references for these.

File: components/org.wso2.carbon.identity.organization.management.organization.user.sharing/src/test/java/org/wso2/carbon/identity/organization/management/organization/user/sharing/UserSharingPolicyHandlerServiceImplTest.java [hunk 9]
```diff
@@ -0,0 +1,469 @@
...
+
+    @AfterMethod
+    public void tearDown() {
+
+        userSharingPolicyHandlerServiceMock.close();
+        dataHolderMock.close();
+        utilsMockedStatic.close();
+    }
+
+    // Tests for GET Shared Orgs.
```
[the remark this answers is not in the document]
Comment (BimsaraBodaragama) [reply components/org.wso2.carbon.identity.organization.management.organization.user.sharing/src/test/java/org/wso2/carbon/identity/organization/management/organization/user/sharing/UserSharingPolicyHandlerServiceImplTest.java]: Addressed in commit: [f32508f](https://github.com/wso2-extensions/identity-organization-management/pull/486/commits/f32508f6f5d94a26897509c5469bdb449b18e7f7)

File: components/org.wso2.carbon.identity.organization.management.organization.user.sharing/src/test/java/org/wso2/carbon/identity/organization/management/organization/user/sharing/UserSharingPolicyHandlerServiceImplTest.java (hunk 5 above)
Comment (BimsaraBodaragama) [reply components/org.wso2.carbon.identity.organization.management.organization.user.sharing/src/test/java/org/wso2/carbon/identity/organization/management/organization/user/sharing/UserSharingPolicyHandlerServiceImplTest.java]: Addressed in commit: [f32508f](https://github.com/wso2-extensions/identity-organization-management/pull/486/commits/f32508f6f5d94a26897509c5469bdb449b18e7f7)

File: components/org.wso2.carbon.identity.organization.management.organization.user.sharing/src/test/java/org/wso2/carbon/identity/organization/management/organization/user/sharing/UserSharingPolicyHandlerServiceImplTest.java (hunk 6 above)
Comment (BimsaraBodaragama) [reply components/org.wso2.carbon.identity.organization.management.organization.user.sharing/src/test/java/org/wso2/carbon/identity/organization/management/organization/user/sharing/UserSharingPolicyHandlerServiceImplTest.java]: Addressed in commit: [f32508f](https://github.com/wso2-extensions/identity-organization-management/pull/486/commits/f32508f6f5d94a26897509c5469bdb449b18e7f7)

File: components/org.wso2.carbon.identity.organization.management.organization.user.sharing/src/test/java/org/wso2/carbon/identity/organization/management/organization/user/sharing/UserSharingPolicyHandlerServiceImplTest.java (hunk 7 above)
Comment (BimsaraBodaragama) [reply components/org.wso2.carbon.identity.organization.management.organization.user.sharing/src/test/java/org/wso2/carbon/identity/organization/management/organization/user/sharing/UserSharingPolicyHandlerServiceImplTest.java]: Addressed in commit: [f32508f](https://github.com/wso2-extensions/identity-organization-management/pull/486/commits/f32508f6f5d94a26897509c5469bdb449b18e7f7)

File: components/org.wso2.carbon.identity.organization.management.organization.user.sharing/src/test/java/org/wso2/carbon/identity/organization/management/organization/user/sharing/constant/TestUserSharingConstants.java (hunk 8 above)
Comment (BimsaraBodaragama) [reply components/org.wso2.carbon.identity.organization.management.organization.user.sharing/src/test/java/org/wso2/carbon/identity/organization/management/organization/user/sharing/constant/TestUserSharingConstants.java]: Addressed in commit: [f32508f](https://github.com/wso2-extensions/identity-organization-management/pull/486/commits/f32508f6f5d94a26897509c5469bdb449b18e7f7)

File: components/org.wso2.carbon.identity.organization.management.organization.user.sharing/src/test/java/org/wso2/carbon/identity/organization/management/organization/user/sharing/UserSharingPolicyHandlerServiceImplTest.java [hunk 10]
```diff
@@ -0,0 +1,445 @@
...
+        OrganizationUserSharingDataHolder dataHolder = mock(OrganizationUserSharingDataHolder.class);
+        when(OrganizationUserSharingDataHolder.getInstance()).thenReturn(dataHolder);
+        OrganizationUserSharingService mockOrgUserSharingService = mock(OrganizationUserSharingService.class);
+        when(dataHolder.getOrganizationUserSharingService()).thenReturn(mockOrgUserSharingService);
+
+        when(mockOrgUserSharingService.getUserAssociationsOfGivenUser(anyString(), anyString())).thenThrow(
+                new OrganizationManagementException(VALIDATE_MSG_EXCEPTION));
+
+        // Invoke the method (expected to throw an exception).
+        userSharingPolicyHandlerService.getSharedOrganizationsOfUser(USER_1_ID, null, null, null, null, false);
```
Comment (SujanSanjula96) [inline components/org.wso2.carbon.identity.organization.management.organization.user.sharing/src/test/java/org/wso2/carbon/identity/organization/management/organization/user/sharing/UserSharingPolicyHandlerServiceImplTest.java:207]: how we are going to validate whether exception is thrown?

Comment (SujanSanjula96) [reply components/org.wso2.carbon.identity.organization.management.organization.user.sharing/src/test/java/org/wso2/carbon/identity/organization/management/organization/user/sharing/UserSharingPolicyHandlerServiceImplTest.java:207]: Let's add `assertThrows` here

File: components/org.wso2.carbon.identity.organization.management.organization.user.sharing/src/test/java/org/wso2/carbon/identity/organization/management/organization/user/sharing/UserSharingPolicyHandlerServiceImplTest.java [hunk 11]
```diff
@@ -0,0 +1,445 @@
...
+                new OrganizationManagementException(VALIDATE_MSG_EXCEPTION));
+
+        // Invoke the method (expected to throw an exception).
+        userSharingPolicyHandlerService.getRolesSharedWithUserInOrganization(USER_1_ID, ORG_1_ID, null, null, null,
+                null, false);
+    }
+
+    // Test case Builders.
+
+    private Map<String, UserAssociation> setExpectedResultsForGetSharedOrgsTestCase1() {
```
Comment (SujanSanjula96) [inline components/org.wso2.carbon.identity.organization.management.organization.user.sharing/src/test/java/org/wso2/carbon/identity/organization/management/organization/user/sharing/UserSharingPolicyHandlerServiceImplTest.java:346]: Here we need to more specific about what is the test case also. At least let's add a comment in the data provider for each case

File: components/org.wso2.carbon.identity.organization.management.organization.user.sharing/src/test/java/org/wso2/carbon/identity/organization/management/organization/user/sharing/UserSharingPolicyHandlerServiceImplTest.java (hunk 10 above)
Comment (BimsaraBodaragama) [reply components/org.wso2.carbon.identity.organization.management.organization.user.sharing/src/test/java/org/wso2/carbon/identity/organization/management/organization/user/sharing/UserSharingPolicyHandlerServiceImplTest.java:207]: We are validating it via
```
 @Test(expectedExceptions = UserSharingMgtClientException.class)
```
in the test declaration.

File: components/org.wso2.carbon.identity.organization.management.organization.user.sharing/src/test/java/org/wso2/carbon/identity/organization/management/organization/user/sharing/UserSharingPolicyHandlerServiceImplTest.java (hunk 11 above)
Comment (BimsaraBodaragama) [reply components/org.wso2.carbon.identity.organization.management.organization.user.sharing/src/test/java/org/wso2/carbon/identity/organization/management/organization/user/sharing/UserSharingPolicyHandlerServiceImplTest.java:346]: Addressed in commit: [97a77c3](https://github.com/wso2-extensions/identity-organization-management/pull/486/commits/97a77c3d194a815e192e279fdbc3d5e1ce7dd11f)

File: components/org.wso2.carbon.identity.organization.management.organization.user.sharing/src/test/java/org/wso2/carbon/identity/organization/management/organization/user/sharing/UserSharingPolicyHandlerServiceImplTest.java (hunk 10 above)
Comment (BimsaraBodaragama) [reply components/org.wso2.carbon.identity.organization.management.organization.user.sharing/src/test/java/org/wso2/carbon/identity/organization/management/organization/user/sharing/UserSharingPolicyHandlerServiceImplTest.java]: Addressed in commit: [0e6ddf8](https://github.com/wso2-extensions/identity-organization-management/pull/486/commits/0e6ddf88b2cfceb4ac829ba5b444643a59c6aed7)

**Review by SujanSanjula96 — APPROVED** [review]

File: components/org.wso2.carbon.identity.organization.management.organization.user.sharing/src/test/java/org/wso2/carbon/identity/organization/management/organization/user/sharing/UserSharingPolicyHandlerServiceImplTest.java [hunk 12]
```diff
@@ -0,0 +1,449 @@
...
+    public Object[][] getSharedOrgsDataProvider() {
+
+        return new Object[][]{
+                {USER_1_ID, setExpectedResultsForGetSharedOrgsTestCase1()}, // Having many shared orgs.
+                {USER_2_ID, setExpectedResultsForGetSharedOrgsTestCase2()}, // Having a single shared org.
+                {USER_3_ID, Collections.emptyMap()} // Having no shared orgs.
+        };
+    }
+
+    @Test(dataProvider = "getSharedOrgsDataProvider")
```
Comment (ShanChathusanda93) [inline components/org.wso2.carbon.identity.organization.management.organization.user.sharing/src/test/java/org/wso2/carbon/identity/organization/management/organization/user/sharing/UserSharingPolicyHandlerServiceImplTest.java]:
```suggestion
    @Test(dataProvider = "sharedOrgsDataProvider")
```

File: components/org.wso2.carbon.identity.organization.management.organization.user.sharing/src/test/java/org/wso2/carbon/identity/organization/management/organization/user/sharing/constant/TestUserSharingConstants.java [hunk 13]
```diff
@@ -0,0 +1,81 @@
...
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.wso2.carbon.identity.organization.management.organization.user.sharing.constant;
+
+/**
+ * Constants related to the test cases of User Sharing Policy Handler service.
+ */
+public class TestUserSharingConstants {
```
Comment (sahandilshan) [inline components/org.wso2.carbon.identity.organization.management.organization.user.sharing/src/test/java/org/wso2/carbon/identity/organization/management/organization/user/sharing/constant/TestUserSharingConstants.java]: Let's add a private constructor to prevent creating objects of this class
